sklearn/ensemble/_stacking.py

```diff
@@ -0,0 +1,681 @@
+        self.final_estimator = final_estimator
+        self.cv = cv
+        self.stack_method = stack_method
+        self.n_jobs = n_jobs
+        self.verbose = verbose
+
+    @abstractmethod
+    def _validate_estimators(self):
+        if self.estimators is None or len(self.estimators) == 0:
+            raise ValueError(
+                "Invalid 'estimators' attribute, 'estimators' should be a list"
+                " of (string, estimator) tuples."
+            )
+        names, estimators = zip(*self.estimators)
+        self._validate_names(names)
+        return names, estimators
+
+    def _clone_final_estimator(self, default):
+        if self.final_estimator is not None:
+            self.final_estimator_ = clone(self.final_estimator)
+        else:
+            self.final_estimator_ = clone(default)
+
+    def set_params(self, **params):
+        """Set the parameters for the stacking estimator.
+
+        Valid parameter keys can be listed with `get_params()`.
+
+        Parameters
+        ----------
+        params : keyword arguments
+            Specific parameters using e.g.
+            `set_params(parameter_name=new_value)`. In addition, to setting the
+            parameters of the stacking estimator, the individual estimator of
+            the stacking estimators can also be set, or can be removed by
+            setting them to 'drop'.
+
+        Examples
+        --------
+        # In this example, the RandomForestClassifier is removed
+        clf1 = LogisticRegression()
+        clf2 = RandomForestClassifier()
+        eclf = StackingClassifier(estimators=[('lr', clf1), ('rf', clf2)]
+        eclf.set_params(rf='drop')
+        """
+        super()._set_params('estimators', **params)
+        return self
+
+    def get_params(self, deep=True):
+        """Get the parameters of the stacking estimator.
+
+        Parameters
+        ----------
+        deep : bool
+            Setting it to True gets the various classifiers and the parameters
+            of the classifiers as well.
+        """
+        return super()._get_params('estimators', deep=deep)
+
+    def _concatenate_predictions(self, predictions):
+        """Concatenate the predictions of each first layer learner.
+
+        This helper is in charge of ensuring the preditions are 2D arrays and
+        it will drop one of the probability column when using probabilities.
+        Indeed, the p(y|c=0) = 1 - sum_{k=1}^{K} p(y|c=k)
+        """
+        X_meta = []
+        for est_idx, preds in enumerate(predictions):
+            # case where the the estimator returned a 1D array
+            if preds.ndim == 1:
+                X_meta.append(preds.reshape(-1, 1))
+            else:
+                if self.stack_method_[est_idx] == 'predict_proba':
+                    # Remove the first column when using probabilities. It is
+                    # collinear with others.
+                    X_meta.append(preds[:, 1:])
```

Comment:
Is it good to remove the first column by default? (maybe we can add an option later, like mlxtend)
(1) I guess we usually don't remove features by default (e.g., in OneHotEncoder)
(2) I understand that for linear models, the first column is not useful, but for tree-based models, maybe the first columns is useful?

Reply:
The redundance does not bring anything for any model here. I don't think that having the option would be useful for any type of model.

Regarding the OneHotEncoder, we should provide such feature as well. I am using the OridinalEncoder to encode binary categories for this reason.

Reply:
We support ``drop=first`` in OHE, but it's disabled by default.

> I don't think that having the option would be useful for any type of model.

Even for tree-based models?

Reply:
I am not sure that it will be helpful.

Reply:
> I am not sure that it will be helpful.

If so, maybe we should not remove by default? We can have an option to remove later.

I think the first column is useful for tree-based models. Supposing that y is determined by the first column, if we remove the first column, it will be much more difficult for the decision tree to fit the dataset (i.e., depth=2 -> depth=n_categories), sometimes even impossible (e.g., when the dataset is small or when users specify max_depth)

Reply:
Uhm. Keeping in mind the use case of stacking (stack different learner), I am not sure a shallow tree would be relevant in practise. WDYT?

Reply:
> Uhm. Keeping in mind the use case of stacking (stack different learner), I am not sure a shallow tree would be relevant in practise. WDYT?

Sorry I can't understand your comment.

I think the issue here is that, unlike linear models, co-linear features are useful for tree-based models.

Reply:
I think this is not really the case. Tree-based models are immune to collinear (or multi-collinear) features instead of benefiting from them. A threshold will use one of the features in an arbitrary manner.

> Sorry I can't understand your comment.

I meant that a decision tree with low depth will not be used in practice in a stacker.

Reply:
Actually co-linear features can actually be beneficial.

E.g. the true model is `y = x1 + x2` with x1 independent of x2, fitting a tree-based model will require a large number of splits on both x1 and x2 to be accurate enough. Engineering x3 = x1 + x2 explicitly will make it possible for the tree to approximate y from x3 with fewer splits (shallow trees).